src/software/util/design_patterns/generic_factory_test.cpp

```diff
@@ -0,0 +1,55 @@
+#include "software/util/design_patterns/generic_factory.cpp"
+
+#include <gtest/gtest.h>
+
+#include <exception>
+#include <iostream>
+
+// Create and register two test generics with the factory here
+class testGeneric {
+};
+
+class testGenericA : public testGeneric {
+    public:
+        static const std::string name;
+};
+const std::string testGenericA::name = "A";
+
+class testGenericB : public testGeneric {
+public:
+    static const std::string name;
+};
+const std::string testGenericB::name = "B";
+
+static TGenericFactory<std::string, testGeneric, testGenericA> testFactoryA;
+static TGenericFactory<std::string, testGeneric, testGenericB> testFactoryB;
+
+
+TEST(GenericFactoryTest, test_create_generic_with_invalid_name)
+{
+    EXPECT_THROW((GenericFactory<std::string, testGeneric>::createGeneric("_FooBar_")), std::invalid_argument);
+}
+
+TEST(GenericFactoryTest, test_create_generic_with_valid_name)
+{
+    auto backend_ptr = GenericFactory<std::string, testGeneric>::createGeneric("A");
```

Comment:
nit: maybe don't call this `backend_ptr`, something more test-related should help avoid confusion

Reply:
Changed to type_ptr to represent unique_ptr<TypeToCreate>